From b37cf63039f1485f90020a50e1b396258388d460 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:06:29 -0400
Subject: [ethernet] qlcnic: Fix max ring count calculation

Message-id: <1408626422-3941-106-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88653
O-Subject: [PATCH RHEL 7.1 105/138] qlcnic: Fix max ring count calculation
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Shahed Shaikh <shahed.shaikh@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id 7b546842b1e87f5b7929cf15e9cd1ac861b14de3

Do not read max rings count from qlcnic_get_nic_info(). Use driver defined
values for 82xx adapters. In case of 83xx adapters, use minimum of firmware
provided and driver defined values.

Signed-off-by: Shahed Shaikh <shahed.shaikh@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_init.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_init.c
index 3ca3118..ba20c72 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_init.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_init.c
@@ -2139,8 +2139,6 @@ static int qlcnic_83xx_get_nic_configuration(struct qlcnic_adapter *adapter)
  ahw->max_mac_filters = nic_info.max_mac_filters;
  ahw->max_mtu = nic_info.max_mtu;
 
- adapter->max_tx_rings = ahw->max_tx_ques;
- adapter->max_sds_rings = ahw->max_rx_ques;
  /* eSwitch capability indicates vNIC mode.
   * vNIC and SRIOV are mutually exclusive operational modes.
   * If SR-IOV capability is detected, SR-IOV physical function
@@ -2161,6 +2159,7 @@ static int qlcnic_83xx_get_nic_configuration(struct qlcnic_adapter *adapter)
 int qlcnic_83xx_configure_opmode(struct qlcnic_adapter *adapter)
 {
  struct qlcnic_hardware_context *ahw = adapter->ahw;
+ u16 max_sds_rings, max_tx_rings;
  int ret;
 
  ret = qlcnic_83xx_get_nic_configuration(adapter);
@@ -2173,18 +2172,21 @@ int qlcnic_83xx_configure_opmode(struct qlcnic_adapter *adapter)
   if (qlcnic_83xx_config_vnic_opmode(adapter))
    return -EIO;
 
-  adapter->max_sds_rings = QLCNIC_MAX_VNIC_SDS_RINGS;
-  adapter->max_tx_rings = QLCNIC_MAX_VNIC_TX_RINGS;
+  max_sds_rings = QLCNIC_MAX_VNIC_SDS_RINGS;
+  max_tx_rings = QLCNIC_MAX_VNIC_TX_RINGS;
  } else if (ret == QLC_83XX_DEFAULT_OPMODE) {
   ahw->nic_mode = QLCNIC_DEFAULT_MODE;
   adapter->nic_ops->init_driver = qlcnic_83xx_init_default_driver;
   ahw->idc.state_entry = qlcnic_83xx_idc_ready_state_entry;
-  adapter->max_sds_rings = QLCNIC_MAX_SDS_RINGS;
-  adapter->max_tx_rings = QLCNIC_MAX_TX_RINGS;
+  max_sds_rings = QLCNIC_MAX_SDS_RINGS;
+  max_tx_rings = QLCNIC_MAX_TX_RINGS;
  } else {
   return -EIO;
  }
 
+ adapter->max_sds_rings = min(ahw->max_rx_ques, max_sds_rings);
+ adapter->max_tx_rings = min(ahw->max_tx_ques, max_tx_rings);
+
  return 0;
 }
 
diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ctx.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ctx.c
index 64dcbf3..a81ad50 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ctx.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_ctx.c
@@ -883,8 +883,6 @@ int qlcnic_82xx_get_nic_info(struct qlcnic_adapter *adapter,
   npar_info->max_rx_ques = le16_to_cpu(nic_info->max_rx_ques);
   npar_info->capabilities = le32_to_cpu(nic_info->capabilities);
   npar_info->max_mtu = le16_to_cpu(nic_info->max_mtu);
-  adapter->max_tx_rings = npar_info->max_tx_ques;
-  adapter->max_sds_rings = npar_info->max_rx_ques;
  }
 
  qlcnic_free_mbx_args(&cmd);
-- 
1.7.1